Add dockerfile which is based on Ubuntu 22.04

andr-sokolov opened the pull request:
Add dockerfile which is based on Ubuntu 22.04

There is no sudo in the ubuntu:22.04 image, so sudo is moved from
README.ubuntu.bash to README.linux.md. Add packages, which are required to build
and run GPDB, to README.ubuntu.bash. The new Dockerfile.ubuntu file is similar
to arenadata/Dockerfile. The first RUN setups the environment to run
compile_gpdb.bash and contains all changes to the ubuntu:22.04 image, which can
be made without the source code. So this layer will not be rebuilt for every
commit.

Commit:
Add dockerfile which is based on Ubuntu 22.04

There is no sudo in the ubuntu:22.04 image, so sudo is moved from
README.ubuntu.bash to README.linux.md. Add packages, which are required to build
GPDB, to README.ubuntu.bash. The new Dockerfile.ubuntu file is similar to
arenadata/Dockerfile. The first RUN setups the environment to run
compile_gpdb.bash, the last - to run regression tests.

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/76568

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/76657

andr-sokolov commented:
bender build

RekGRpth commented:
> The new Dockerfile.ubuntu file is similar to
> arenadata/Dockerfile.

Why are some steps in a different order?

RekGRpth commented on arenadata/Dockerfile.ubuntu:
```diff
@@ -0,0 +1,57 @@
+
+WORKDIR /home/gpadmin
+
+COPY README.ubuntu.bash ./
+RUN ./README.ubuntu.bash && \
+    rm README.ubuntu.bash && \
```
centos's docker file does not contain these steps

andr-sokolov replied:
The current approach is better than approach used in centos's docker file, because we keep package list to install in one place

RekGRpth commented:
```bash
export GP_MAJOR=6
docker build \
    --build-arg GP_MAJOR=$GP_MAJOR \
    --file "arenadata/Dockerfile.ubuntu" \
    --pull \
    --tag "gpdb${GP_MAJOR}u" \
    . 2>&1 | tee "build${GP_MAJOR}u.log"
...
#13 96.88 /usr/bin/ld: transform.o: relocation R_X86_64_32S against `.rodata' can not be used when making a PIE object; recompile with -fPIE
#13 96.89 /usr/bin/ld: failed to set dynamic section sizes: bad value
#13 96.89 collect2: error: ld returned 1 exit status
#13 96.89 make[4]: *** [Makefile:39: gpfdist] Error 1
#13 96.89 make[3]: *** [Makefile:31: install-gpfdist-recurse] Error 2
#13 96.89 make[3]: *** Waiting for unfinished jobs....
#13 97.31 /usr/bin/ld: pg_backup_tar.o: in function `newTempFile':
#13 97.31 /home/gpadmin/gpdb_src/src/bin/pg_dump/pg_backup_tar.c:506: warning: the use of `tempnam' is dangerous, better use `mkstemp'
#13 98.50 /usr/bin/ld: pg_backup_tar.o: in function `newTempFile':
#13 98.50 /home/gpadmin/gpdb_src/src/bin/pg_dump/pg_backup_tar.c:506: warning: the use of `tempnam' is dangerous, better use `mkstemp'
#13 98.55 make[2]: *** [Makefile:34: install-bin-recurse] Error 2
#13 98.55 make[1]: *** [GNUmakefile:12: install-src-recurse] Error 2
#13 98.55 make: *** [Makefile:318: dist] Error 2
#13 ERROR: process "/bin/sh -c bash /home/gpadmin/gpdb_src/concourse/scripts/compile_gpdb.bash" did not complete successfully: exit code: 2
------
 > [build 3/3] RUN bash /home/gpadmin/gpdb_src/concourse/scripts/compile_gpdb.bash:
96.89 make[4]: *** [Makefile:39: gpfdist] Error 1
96.89 make[3]: *** [Makefile:31: install-gpfdist-recurse] Error 2
96.89 make[3]: *** Waiting for unfinished jobs....
pg_backup_tar.o: in function `newTempFile':
97.31 /home/gpadmin/gpdb_src/src/bin/pg_dump/pg_backup_tar.c:506: warning: the use of `tempnam' is dangerous, better use `mkstemp'
pg_backup_tar.o: in function `newTempFile':
98.50 /home/gpadmin/gpdb_src/src/bin/pg_dump/pg_backup_tar.c:506: warning: the use of `tempnam' is dangerous, better use `mkstemp'
98.55 make[2]: *** [Makefile:34: install-bin-recurse] Error 2
98.55 make[1]: *** [GNUmakefile:12: install-src-recurse] Error 2
98.55 make: *** [Makefile:318: dist] Error 2
------
Dockerfile.ubuntu:29
--------------------
  27 |
  28 |     # Compile with running mocking tests
  29 | >>> RUN bash /home/gpadmin/gpdb_src/concourse/scripts/compile_gpdb.bash
  30 |
  31 |     FROM base as code
--------------------
ERROR: failed to solve: process "/bin/sh -c bash /home/gpadmin/gpdb_src/concourse/scripts/compile_gpdb.bash" did not complete successfully: exit code: 2
```

RekGRpth commented on arenadata/Dockerfile.ubuntu:
```diff
@@ -0,0 +1,57 @@
+WORKDIR /home/gpadmin
+
+COPY README.ubuntu.bash ./
+RUN ./README.ubuntu.bash && \
```
try use modern syntax with set -eux and ;

andr-sokolov replied:
done

RekGRpth commented on arenadata/Dockerfile.ubuntu:
```diff
@@ -0,0 +1,57 @@
+RUN ./README.ubuntu.bash && \
+    rm README.ubuntu.bash && \
+    wget $sigar $sigar_headers $libxerces $libxerces_dev && \
+    apt install ./*.deb && \
```
use -y flag with install

andr-sokolov replied:
done

Commit:
Add -y flag and replace "&&" with "set -eux" and ";"

Commit:
Move WORKDIR

andr-sokolov commented:
> > The new Dockerfile.ubuntu file is similar to
> > arenadata/Dockerfile.
> 
> Why are some steps in a different order?

I've updated the PR description and moved WORDIR to the same place as in the dockerfile for CentOS

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/76851

KnightMurloc commented on arenadata/Dockerfile.ubuntu:
```diff
@@ -0,0 +1,59 @@
+COPY --from=code /home/gpadmin/gpdb_src gpdb_src
+COPY --from=build /home/gpadmin/bin_gpdb /home/gpadmin/bin_gpdb
+
+# Install entab and pg_bsd_indent used by pgindent utility.
+# This should be done using gpdb sources.
+RUN set -eux; \
+    make -C gpdb_src/src/tools/entab install clean; \
+    wget https://ftp.postgresql.org/pub/dev/pg_bsd_indent-1.3.tar.gz -O - | tar -xzf -; \
+    make install -C pg_bsd_indent; \
+    rm -r pg_bsd_indent; \
```
Why is this necessary? We don't do this in a container for centos.

andr-sokolov replied:
https://github.com/arenadata/gpdb/blob/6ad72daf603ca7b061863e6e7354d45e44e24e89/arenadata/Dockerfile#L88C1-L91C25

KnightMurloc commented on arenadata/Dockerfile.ubuntu:
```diff
@@ -0,0 +1,59 @@
+    make install -C pg_bsd_indent; \
+    rm -r pg_bsd_indent; \
+# libipc-run-perl and rsync are needed to run regression tests
+# iproute2 is needed to run GPDB
```
if this package is needed for gpdb to work, why not add it to `README.ubuntu.bash`?

andr-sokolov replied:
done

KnightMurloc commented on arenadata/Dockerfile.ubuntu:
```diff
@@ -0,0 +1,59 @@
+    rm -r pg_bsd_indent; \
+# libipc-run-perl and rsync are needed to run regression tests
+# iproute2 is needed to run GPDB
+    apt install -y libipc-run-perl rsync iproute2; \
+# The en_US.UTF-8 locale is needed to run GPDB
+    locale-gen en_US.UTF-8; \
+# To run sshd directly, but not using `/etc/init.d/ssh start`
+    mkdir /run/sshd; \
+# Alter precedence in favor of IPv4 during resolving
+    echo 'precedence ::ffff:0:0/96  100' >> /etc/gai.conf
```
can we do this before compiling gpdb so that we don't have to do it every time we rebuild?

andr-sokolov replied:
done

KnightMurloc reviewed (commented):
There are not enough packages to run the behave tests. at least the behave package itself. There also a java package missing, which I understand is needed for pxf tests, or will pxf be tested in another container? There are also no more java related steps https://github.com/arenadata/gpdb/blob/6ad72daf603ca7b061863e6e7354d45e44e24e89/arenadata/Dockerfile#L80-L81.

andr-sokolov commented:
> There are not enough packages to run the behave tests. at least the behave package itself. There also a java package missing, which I understand is needed for pxf tests, or will pxf be tested in another container? There are also no more java related steps
> 
> https://github.com/arenadata/gpdb/blob/6ad72daf603ca7b061863e6e7354d45e44e24e89/arenadata/Dockerfile#L80-L81
> .

This image version purpose is to successfully execute compile_gpdb.bash and run compiled GPDB. Compatibility with all the tests will be checked and implemented on the next steps

Commit:
Fixes after review

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/76944

BenderArenadata commented:
Failed job Behave tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1736023

Commit:
Merge branch 'adb-6.x-dev' into ADBDEV-6043

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/76989
